beacon-chain/sync/fork_watcher.go

```diff
@@ -0,0 +1,108 @@
+package sync
+
+import (
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+	"github.com/prysmaticlabs/prysm/shared/p2putils"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/slotutil"
+)
+
+// Is a background routine that observes for new incoming forks. Depending on the epoch
+// it will be in charge of subscribing/unsubscribing the relevant topics at the fork boundaries.
+func (s *Service) forkWatcher() {
+	slotTicker := slotutil.NewSlotTicker(s.cfg.Chain.GenesisTime(), params.BeaconConfig().SecondsPerSlot)
+	for {
+		select {
+		// In the event of a node restart, we will still end up subscribing to the correct
+		// topics during/after the fork epoch. This routine is to ensure correct
+		// subscriptions for nodes running before a fork epoch.
+		case currSlot := <-slotTicker.C():
+			currEpoch := core.SlotToEpoch(currSlot)
+			if err := s.checkForNextEpochFork(currEpoch); err != nil {
+				log.WithError(err).Error("Unable to check for fork in the next epoch")
+				continue
+			}
+			if err := s.checkForPreviousEpochFork(currEpoch); err != nil {
+				log.WithError(err).Error("Unable to check for fork in the previous epoch")
+				continue
+			}
+		case <-s.ctx.Done():
+			log.Debug("Context closed, exiting goroutine")
+			slotTicker.Done()
+			return
+		}
+	}
+}
+
+// Checks if there is a fork in the next epoch.
+func (s *Service) checkForNextEpochFork(currEpoch types.Epoch) error {
+	genRoot := s.cfg.Chain.GenesisValidatorRoot()
+	isNextForkEpoch, err := p2putils.IsForkNextEpoch(s.cfg.Chain.GenesisTime(), genRoot[:])
+	if err != nil {
+		return errors.Wrap(err, "Could not retrieve next fork epoch")
+	}
+	// In preparation for the upcoming fork
+	// in the following epoch, the node
+	// will subscribe the new topics in advance.
+	if isNextForkEpoch {
+		nextEpoch := currEpoch + 1
+		if nextEpoch == params.BeaconConfig().AltairForkEpoch {
+			digest, err := p2putils.ForkDigestFromEpoch(nextEpoch, genRoot[:])
+			if err != nil {
+				return errors.Wrap(err, "Could not retrieve fork digest")
+			}
+			if s.subHandler.digestExists(digest) {
+				return nil
+			}
+			s.registerSubscribers(nextEpoch, digest)
+			s.registerRPCHandlersAltair()
+		}
+	}
+	return nil
+}
+
+// Checks if there is a fork in the previous epoch.
+func (s *Service) checkForPreviousEpochFork(currEpoch types.Epoch) error {
```

Comment:
Can you elaborate on the commentary here? This function does a lot more than check if the previous epoch was a fork.

Reply:
agreed, the multiple responsibilities can be split up